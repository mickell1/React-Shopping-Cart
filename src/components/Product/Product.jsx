import React, { Component } from "react";
import PropTypes from 'prop-types'
import Counter from "../Counter/Counter";
import Button from '../Button/Button'
import './Product.scss'

class Product extends Component {
  constructor(props) {
    super(props);
    this.state = {
      selectedProduct: {},
      quickViewProduct: {},
      isAdded: false
    };
  }
  addToBasket(image, title, price, productId, quantity) {
    this.setState(
      {
        selectedProduct: {
          image: image,
          title: title,
          price: price,
          productId: productId,
          quantity: quantity
        }
      },
      () => {
        this.props.addToBasket(this.state.selectedProduct);
      }
    );
    this.setState(
      {
        isAdded: true
      },
      () => {
        setTimeout(() => {
          this.setState({
            isAdded: false,
            selectedProduct: {}
          });
        }, 3500);
      }
    );
  }
  quickView(image, title, price, productId, quantity) {
    this.setState(
      {
        quickViewProduct: {
          image: image,
          title: title,
          price: price,
          productId: productId,
          quantity: quantity,
        }
      },
      () => {
        this.props.openModal(this.state.quickViewProduct);
      }
    );
  }
  render() {
    let image = this.props.image;
    let title = this.props.title;
    let price = this.props.price;
    let productId = this.props.productId;
    let quantity = this.props.productQuantity;
    return (
      <div className="product">
        <div className="product-image">
          <img
            src={this.props.image}
            alt={this.props.title}
            onClick={this.quickView.bind(
              this,
              image,
              title,
              price,
              productId,
              quantity
            )}
          />
        </div>
        <h4 className="product-name">{this.props.title}</h4>
        <p className="product-price">{this.props.price}</p>
        <Counter
          productQuantity={quantity}
          updateQuantity={this.props.updateQuantity}
          resetQuantity={this.resetQuantity}
        />
        <div className="product-action">
          <Button
            className={!this.state.isAdded ? "" : "added"}
            onClick={this.addToBasket.bind(
              this,
              image,
              title,
              price,
              productId,
              quantity
            )}
          >
            {!this.state.isAdded ? "ADD TO CART" : "âœ” ADDED"}
          </Button>
        </div>
      </div>
    );
  }
}

Product.propTypes = {
  product: PropTypes.shape({
  productId: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
  title: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
  }),
}

export default Product;
