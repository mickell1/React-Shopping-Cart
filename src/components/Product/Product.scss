@import "../../variables.scss";

.product { 
  display: flex;
  flex-direction: column;
  border: 1px solid $color-yellowy-grey;
  width: 220px;
  padding: 15px;
  justify-content: space-between;
  background-color: $color-white;

  .product-image {
    overflow: hidden;
    border-radius: 2px 2px 0 0;
    max-height: 200px;
    img {
      cursor: zoom-in;
      width: 100%;
      height: auto;
      transition: transform 300ms ease-in;
      transform: scale(1);
      @media (min-width: 991px) {
        min-height: 200px;
      }
      &:hover {
        transform: scale(1.1);
      }
    }
  }
  
.product-price {
  margin-bottom: 0;
  color: $color-black;
  padding: 0 16px;
  text-align: center;
  &:before {
    content: "Â£ ";
  }
}

  .product-action {
    padding: 16px;
    button {
      width: 100%;
      transition: all 300ms ease-in;
      &.added {
        background: $color-light-green;
      }
    }
  }
  &:hover {
    box-shadow: $color-blue;
  }
}

.step-input {
  display: flex;
  display: -webkit-flex;
  color: $color-gray-medium;
  max-width: 120px;
  margin: 0 auto;
  .increase,
  .decrease {
    height: 24px;
    width: 24px;
    border: 1px solid $color-gray-dark;
    text-align: center;
    box-sizing: border-box;
    border-radius: 50%;
    text-decoration: none;
    color: inherit;
    font-size: 24px;
    line-height: 22px;
    -moz-user-select: none;
    -webkit-user-select: none;
    &:hover {
      color: $color-red;
      border-color: $color-red;
    }
    &:active {
      color: $color-white;
      border-color: $color-red;
      background: lighten($color-red, 15%);
    }
  }
  .quantity {
    height: 24px;
    width: 48px;
    text-align: center;
    margin: 0 12px;
    border-radius: 2px;
    border: 1px solid $color-gray-dark;
    &:focus {
      outline: none;
      border-color: $color-light-green;
    }
  }
}
