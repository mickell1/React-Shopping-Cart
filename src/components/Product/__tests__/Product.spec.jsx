import React from 'react';
import { shallow } from 'enzyme';
import Product from '../Product';

describe('Product', () => {
  let wrapper;

  const defaultProps = {
    image: 'Iphone',
    title: 'Iphone x',
    price: 800,
    productId: 1,
    quantity: 1,
    updateQuantity: jest.fn(),
    addToBasket: jest.fn()
  }

  beforeEach(() => {
    wrapper = shallow(<Product {...defaultProps} /> )  
  })

  it('should render the Product Component correctly', () => {   
    expect(wrapper).toMatchSnapshot();
  });

  it('should call quickView', () => {
    const props = { ...defaultProps, openModal: jest.fn(), quickView: jest.fn() }

     wrapper = shallow(
      <Product {...props} />
    );

    const image = wrapper.find('img').first();
    image.simulate('click');

    expect(props.openModal).toHaveBeenCalled();
  });

  it('should call addToBasket', () => {
    const props = { ...defaultProps, addToBasket: jest.fn() }

     wrapper = shallow(
      <Product {...props} />
    );

    const button = wrapper.find('Button').first();
    button.simulate('click');

    expect(props.addToBasket).toHaveBeenCalled();
    expect(wrapper.find('div').first().hasClass('product')).toEqual(true);
  });
});