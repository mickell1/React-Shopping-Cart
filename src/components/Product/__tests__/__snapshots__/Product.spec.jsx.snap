// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product should render the Product Component correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Product
    addToBasket={[MockFunction]}
    image="Iphone"
    price={800}
    productId={1}
    quantity={1}
    title="Iphone x"
    updateQuantity={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="product-image"
        >
          <img
            alt="Iphone x"
            onClick={[Function]}
            src="Iphone"
          />
        </div>,
        <h4
          className="product-name"
        >
          Iphone x
        </h4>,
        <p
          className="product-price"
        >
          800
        </p>,
        <Counter
          updateQuantity={[MockFunction]}
        />,
        <div
          className="product-action"
        >
          <Button
            className=""
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            ADD TO CART
          </Button>
        </div>,
      ],
      "className": "product",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <img
            alt="Iphone x"
            onClick={[Function]}
            src="Iphone"
          />,
          "className": "product-image",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "alt": "Iphone x",
            "onClick": [Function],
            "src": "Iphone",
          },
          "ref": null,
          "rendered": null,
          "type": "img",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Iphone x",
          "className": "product-name",
        },
        "ref": null,
        "rendered": "Iphone x",
        "type": "h4",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": 800,
          "className": "product-price",
        },
        "ref": null,
        "rendered": 800,
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "productQuantity": undefined,
          "resetQuantity": undefined,
          "updateQuantity": [MockFunction],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Button
            className=""
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            ADD TO CART
          </Button>,
          "className": "product-action",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "ADD TO CART",
            "className": "",
            "disabled": false,
            "onClick": [Function],
            "type": "button",
          },
          "ref": null,
          "rendered": "ADD TO CART",
          "type": [Function],
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="product-image"
          >
            <img
              alt="Iphone x"
              onClick={[Function]}
              src="Iphone"
            />
          </div>,
          <h4
            className="product-name"
          >
            Iphone x
          </h4>,
          <p
            className="product-price"
          >
            800
          </p>,
          <Counter
            updateQuantity={[MockFunction]}
          />,
          <div
            className="product-action"
          >
            <Button
              className=""
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              ADD TO CART
            </Button>
          </div>,
        ],
        "className": "product",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <img
              alt="Iphone x"
              onClick={[Function]}
              src="Iphone"
            />,
            "className": "product-image",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "alt": "Iphone x",
              "onClick": [Function],
              "src": "Iphone",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Iphone x",
            "className": "product-name",
          },
          "ref": null,
          "rendered": "Iphone x",
          "type": "h4",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": 800,
            "className": "product-price",
          },
          "ref": null,
          "rendered": 800,
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "productQuantity": undefined,
            "resetQuantity": undefined,
            "updateQuantity": [MockFunction],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Button
              className=""
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              ADD TO CART
            </Button>,
            "className": "product-action",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "ADD TO CART",
              "className": "",
              "disabled": false,
              "onClick": [Function],
              "type": "button",
            },
            "ref": null,
            "rendered": "ADD TO CART",
            "type": [Function],
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
